components/feature/downloads/src/main/java/mozilla/components/feature/downloads/manager/FetchDownloadManager.kt

```diff
@@ -0,0 +1,168 @@
+package mozilla.components.feature.downloads.manager
+
+import android.Manifest.permission.INTERNET
+import android.Manifest.permission.WRITE_EXTERNAL_STORAGE
+import android.annotation.TargetApi
+import android.content.Context
+import android.net.Uri
+import android.os.Build
+import android.os.Build.VERSION.SDK_INT
+import android.os.Environment
+import androidx.core.content.getSystemService
+import androidx.core.net.toUri
+import kotlinx.coroutines.Dispatchers.IO
+import kotlinx.coroutines.withContext
+import mozilla.components.browser.session.Download
+import mozilla.components.concept.fetch.Client
+import mozilla.components.concept.fetch.Headers.Names.CONTENT_LENGTH
+import mozilla.components.concept.fetch.Headers.Names.CONTENT_TYPE
+import mozilla.components.concept.fetch.MutableHeaders
+import mozilla.components.concept.fetch.Request
+import mozilla.components.concept.fetch.Response
+import mozilla.components.support.ktx.android.content.isPermissionGranted
+import java.io.File
+import java.io.FileOutputStream
+import java.io.OutputStream
+
+class FetchDownloadManager(
+    private val applicationContext: Context,
+    private val httpClient: Client
+) : DownloadManager {
+
+    override suspend fun download(download: Download, refererUrl: String, cookie: String): CompletedDownload? {
+        if (download.isSupportedProtocol()) {
+            // We are ignoring everything that is not http or https. This is a limitation of
+            // GeckoView: https://bugzilla.mozilla.org/show_bug.cgi?id=1501735 and
+            // https://bugzilla.mozilla.org/show_bug.cgi?id=1432949
+            return null
+        }
+
+        if (!applicationContext.isPermissionGranted(INTERNET, WRITE_EXTERNAL_STORAGE)) {
+            throw SecurityException("You must be granted INTERNET and WRITE_EXTERNAL_STORAGE permissions")
+        }
+
+        val headers = MutableHeaders().apply {
+            safeAppend(CONTENT_TYPE, download.contentType)
+            safeAppend("User-Agent", download.userAgent)
+            safeAppend("Cookie", cookie)
+            safeAppend("Referer", refererUrl)
+        }
+        val request = Request(url = download.url, headers = headers)
+
+        return startDownload(download, request, refererUrl.toUri())
+    }
+
+    private suspend fun startDownload(download: Download, request: Request, refererUrl: Uri) = withContext(IO) {
+        val response = httpClient.fetch(request)
```

Comment:
Ideally we wouldn't perform the download here immediately. Instead we should hand it off to a foreground(!) service, primarily to (a) download independently from the app task being still there (b) continuing to download without getting killed. Using a service should also make it easier to use a richer "ongoing download" notification later on - that allows the user to pause/cancel etc.